openstack/pkg/designate/client.go

```diff
@@ -0,0 +1,213 @@
+package designate
+
+import (
+	"errors"
+	"fmt"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/dns/v2/recordsets"
+	"github.com/gophercloud/gophercloud/openstack/dns/v2/zones"
+	"github.com/libdns/libdns"
+	"os"
+	"time"
+)
+
+func (p *Provider) getRecords(recordSets []recordsets.RecordSet) ([]libdns.Record, error) {
+	var records []libdns.Record
+	for _, j := range recordSets {
+		tmp := libdns.Record{
+			ID:   j.ID,
+			Type: j.Type,
+			Name: j.Name,
+			TTL:  time.Duration(j.TTL) * time.Second,
+		}
+		records = append(records, tmp)
+	}
+
+	return records, nil
+}
+
+func (p *Provider) getRecordID(recordName string, zone string) (string, error) {
+	recordName = recordName + zone
+	listOpts := recordsets.ListOpts{
+		Type: "TXT",
+	}
+
+	allPages, err := recordsets.ListByZone(p.dnsClient, p.zoneID, listOpts).AllPages()
+	if err != nil {
+		return "", err
+	}
+
+	allRecordSets, err := recordsets.ExtractRecordSets(allPages)
+	if err != nil {
+		return "", err
+	}
+
+	for _, rr := range allRecordSets {
+		if recordName == rr.Name {
+			return rr.ID, nil
+		}
+	}
+
+	return "", nil
+}
+
+func (p *Provider) setRecordID(recordID string) {
+	p.recordID = recordID
+}
+
+func (p *Provider) setDescription(desc string) error {
+	if desc == "" {
+		p.dnsDescription = "example description"
+	}
+	p.dnsDescription = desc
+
+	return nil
+}
+
+func (p *Provider) createRecord(record libdns.Record, zone string) error {
+	createOpts := recordsets.CreateOpts{
+		Name:    record.Name + zone,
+		Type:    record.Type,
+		TTL:     int(record.TTL / time.Second),
+		Records: []string{record.Value},
+	}
+
+	exist, err := p.getRecordID(record.Name, zone)
+	if err != nil {
+		return fmt.Errorf("cannot get recordID: %v", err)
+	}
+
+	if exist != "" {
+		return errors.New("DNS record already exist")
+	}
+
+	_, err = recordsets.Create(p.dnsClient, p.zoneID, createOpts).Extract()
+	if err != nil {
+		return fmt.Errorf("cannot create DNS record: %v", err)
+	}
+
+	return nil
+}
+
+func (p *Provider) updateRecord(record libdns.Record, zone string) error {
+	updateOpts := recordsets.UpdateOpts{
+		TTL:     intToPointer(int(record.TTL / time.Second)),
+		Records: []string{record.Value},
+	}
+
+	// Update updates a recordset in a given zone
+	_, err := recordsets.Update(p.dnsClient, p.zoneID, p.recordID, updateOpts).Extract()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Provider) deleteRecord(record libdns.Record, zone string) error {
+	err := recordsets.Delete(p.dnsClient, p.zoneID, p.recordID).ExtractErr()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Provider) exportEnvVariables() error {
```

Comment:
Does the underlying library _require_ env variables for configuration? If so, yuck; you'll need to make all uses of it mutually exclusive so that multiple configs can run on top of each other. It'll be very inefficient but there's no other way to do it correctly if that's the case. (They should fix that upstream.)

Reply:
Unfortunately from what I see they require env variables for the auth configuration :-(

Reply:
Hmmm. I dunno if that's gonna fly. They don't expose _any_ way to manually configure with code?

Reply:
I think I found a way :-)